src/evaluators/BinaryExpression.js

```diff
@@ -175,19 +187,16 @@ export function computeBinary(
 
   if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
     // generate error if binary operation might throw or have side effects
+    getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
     if (realm.isInPureScope()) {
-      realm.evaluateWithPossibleThrowCompletion(
-        () => {
-          getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-          return realm.intrinsics.undefined;
-        },
+      return realm.evaluateWithPossibleThrowCompletion(
+        () => AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc, undefined, true),
```

Comment:
Why do you want a temporal value here? By definition, nothing this operation does may affect the state of Prepack so it really should not matter when this happens.

Reply:
It may affect the state of leaked objects so when it is invoked matters.

Reply:
It also needs to track when it might throw so that we can later wrap the operation in a try/catch.

The problem is that we don't *always* need this to be temporal. Only if the arguments has some unknown `valueOf` semantic or similar side-effect. We should be able to communicate that from getPureBinaryOperationResultType.